File: src/main/java/com/zeros/scriptassistant/Device.java
package com.zeros.scriptassistant;

import cn.hutool.core.io.FileUtil;
import cn.hutool.core.thread.ThreadUtil;
import org.w3c.dom.Document;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;

import javax.swing.tree.DefaultMutableTreeNode;
import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import java.awt.*;
import java.io.ByteArrayInputStream;
import java.io.File;
import java.io.InputStream;
import java.util.*;

public class Device {
    public static final String OBJECT_NAME = "d";
    public String serial;
    public U2 u2;
    public String dumpSaveDir;
    public byte[] screenshot;
    public String hierarchy;
    public Document hierarchyDoc;
    public DefaultMutableTreeNode treeNode;
    public Set<Rectangle> allNodeRect = new HashSet<>();
    public Map<String, Map<String, Integer>> attributeMap;
    private static long lastScreenshotModifyTime;
    private static long lastHierarchyModifyTime;
    private File screenShotSaveFile;
    private File hierarchySaveFile;
    private String alias = "d";
    private DeviceAliasConfig deviceAliasConfig;

    public Device(String serial, DeviceAliasConfig deviceAliasConfig) {
        this.serial = serial;
        u2 = new U2();
        this.deviceAliasConfig = deviceAliasConfig;
        alias = deviceAliasConfig.getAlias(serial);
    }

    public boolean init(String pythonPath, String saveDir) {
        dumpSaveDir = saveDir;
        screenShotSaveFile = new File(dumpSaveDir, "screen.png");
        hierarchySaveFile = new File(dumpSaveDir, "hierarchy.xml");
        boolean success = u2.init(pythonPath, serial, OBJECT_NAME);
        if (!success) {
            u2.destroy();
        }
        return success;
    }

    public String getSerial() {
        return serial;
    }

    public byte[] getScreenshot() {
        for (int i = 0; i < 50; i++) {
            ThreadUtil.sleep(100);
            Date date = FileUtil.lastModifiedTime(screenShotSaveFile);
            if (date != null && date.getTime() != lastScreenshotModifyTime) {
                lastScreenshotModifyTime = date.getTime();
                break;
            }
        }
        screenshot = FileUtil.readBytes(screenShotSaveFile);
        return screenshot;
    }

    public DefaultMutableTreeNode getHierarchy() {
        for (int i = 0; i < 50; i++) {
            ThreadUtil.sleep(100);
            Date date = FileUtil.lastModifiedTime(hierarchySaveFile);
            if (date != null && date.getTime() != lastHierarchyModifyTime) {
                lastHierarchyModifyTime = date.getTime();
                break;
            }
        }
        hierarchy = FileUtil.readUtf8String(hierarchySaveFile);
        DocumentBuilderFactory dbFactory = DocumentBuilderFactory.newInstance();
        DocumentBuilder dBuilder = null;
        try {
            dBuilder = dbFactory.newDocumentBuilder();
            InputStream is = new ByteArrayInputStream(hierarchy.getBytes("UTF-8"));
            Document doc = dBuilder.parse(is);
            doc.getDocumentElement().normalize();
            hierarchyDoc = doc;
            NodeList nodeList = doc.getDocumentElement().getChildNodes();
            treeNode = new DefaultMutableTreeNode(new NodeInfo(doc.getDocumentElement().getNodeName()));
            addNodes(nodeList, treeNode);
            allNodeRect.clear();
            obtainAllNodeRect(treeNode);
            obtainAttributeMap();
            return treeNode;
        } catch (Exception e) {
            e.printStackTrace();
        }
        return null;
    }

    private void addNodes(NodeList nList, DefaultMutableTreeNode parent) {
        for (int i = 0; i < nList.getLength(); i++) {
            Node node = nList.item(i);
            if (node.getNodeType() == Node.ELEMENT_NODE) {
                DefaultMutableTreeNode child = new DefaultMutableTreeNode(new NodeInfo(node));
                parent.add(child);
                // Recursive call for the current node's children
                if (node.hasChildNodes()) {
                    addNodes(node.getChildNodes(), child);
                }
            }
        }
    }

    public void obtainAllNodeRect(DefaultMutableTreeNode root) {
        NodeInfo nodeInfo = (NodeInfo) root.getUserObject();
        allNodeRect.add(nodeInfo.bounds);
        int childCount = root.getChildCount();
        for (int i = 0; i < childCount; i++) {
            DefaultMutableTreeNode child = (DefaultMutableTreeNode) root.getChildAt(i);
            obtainAllNodeRect(child);
        }
    }

    public void obtainAttributeMap() {
        NodeList node = hierarchyDoc.getDocumentElement().getElementsByTagName("node");
        attributeMap = new HashMap<>();
        for (int i = 0; i < node.getLength(); i++) {
            Node item = node.item(i);
            for (String prop : XPathLite.props) {
                String nodeValue = item.getAttributes().getNamedItem(prop).getNodeValue();
                if (nodeValue != null && !nodeValue.isEmpty()) {
                    Map<String, Integer> propMap = attributeMap.computeIfAbsent(prop, k -> new HashMap<>());
                    Integer count = propMap.get(nodeValue);
                    if (count == null) {
                        count = 0;
                    }
                    propMap.put(nodeValue, count + 1);
                }
            }
        }
    }

    public void dumpUI() {
        u2.dumpUI(screenShotSaveFile, hierarchySaveFile);
    }

    public void close() {
        u2.destroy();
    }

    public void setAlias(String alias) {
        this.alias = alias;
        // 保存 alias
        deviceAliasConfig.setAlias(serial, alias);
    }

    public void setAlias(int index) {
        if (!alias.equals("d")) {
            return;
        }
        if (index > 0) {
            alias = "d" + (index + 1);
        } else {
            alias = "d";
        }
        // 保存 alias
        deviceAliasConfig.setAlias(serial, alias);
    }

    public String getAlias() {
        return alias;
    }
}

File: src/main/kotlin/com/zeros/scriptassistant/Device.kt
package com.zeros.scriptassistant

import cn.hutool.core.io.FileUtil
import cn.hutool.core.thread.ThreadUtil
import org.w3c.dom.Document
import org.w3c.dom.Node
import org.w3c.dom.NodeList
import java.awt.Rectangle
import java.io.ByteArrayInputStream
import java.io.File
import java.util.*
import javax.swing.tree.DefaultMutableTreeNode
import javax.xml.parsers.DocumentBuilderFactory

class Device(val serial: String, private val deviceAliasConfig: DeviceAliasConfig) {
    companion object {
        const val OBJECT_NAME = "d"
    }

    var u2 = U2()
    var dumpSaveDir: String? = null
    private var screenshotBytes: ByteArray? = null
    var hierarchy: String? = null
    var hierarchyDoc: Document? = null
    var treeNode: DefaultMutableTreeNode? = null
    val allNodeRect = HashSet<Rectangle>()
    var attributeMap: MutableMap<String, MutableMap<String, Int>>? = null

    private var lastScreenshotModifyTime: Long = 0
    private var lastHierarchyModifyTime: Long = 0
    private lateinit var screenShotSaveFile: File
    private lateinit var hierarchySaveFile: File
    var alias: String = "d"

    fun init(pythonPath: String, saveDir: String): Boolean {
        dumpSaveDir = saveDir
        screenShotSaveFile = File(dumpSaveDir, "screen.png")
        hierarchySaveFile = File(dumpSaveDir, "hierarchy.xml")
        alias = deviceAliasConfig.getAlias(serial) ?: "d"

        val success = u2.init(pythonPath, serial, OBJECT_NAME)
        if (!success) {
            u2.destroy()
        }
        return success
    }

    fun getScreenshot(): ByteArray? {
        for (i in 0 until 50) {
            ThreadUtil.sleep(100)
            val date = FileUtil.lastModifiedTime(screenShotSaveFile)
            if (date != null && date.time != lastScreenshotModifyTime) {
                lastScreenshotModifyTime = date.time
                break
            }
        }
        screenshotBytes = FileUtil.readBytes(screenShotSaveFile)
        return screenshotBytes
    }

    fun getHierarchy(): DefaultMutableTreeNode? {
        for (i in 0 until 50) {
            ThreadUtil.sleep(100)
            val date = FileUtil.lastModifiedTime(hierarchySaveFile)
            if (date != null && date.time != lastHierarchyModifyTime) {
                lastHierarchyModifyTime = date.time
                break
            }
        }
        hierarchy = FileUtil.readUtf8String(hierarchySaveFile)

        return try {
            val dbFactory = DocumentBuilderFactory.newInstance()
            val dBuilder = dbFactory.newDocumentBuilder()
            val `is` = ByteArrayInputStream(hierarchy!!.toByteArray(Charsets.UTF_8))
            val doc = dBuilder.parse(`is`)
            doc.documentElement.normalize()
            hierarchyDoc = doc

            val nodeList = doc.documentElement.childNodes
            treeNode = DefaultMutableTreeNode(NodeInfo(doc.documentElement.nodeName))
            addNodes(nodeList, treeNode!!)
            allNodeRect.clear()
            obtainAllNodeRect(treeNode!!)
            obtainAttributeMap()
            treeNode
        } catch (e: Exception) {
            e.printStackTrace()
            null
        }
    }

    private fun addNodes(nList: NodeList, parent: DefaultMutableTreeNode) {
        for (i in 0 until nList.length) {
            val node = nList.item(i)
            if (node.nodeType == Node.ELEMENT_NODE) {
                val child = DefaultMutableTreeNode(NodeInfo(node))
                parent.add(child)
                if (node.hasChildNodes()) {
                    addNodes(node.childNodes, child)
                }
            }
        }
    }

    private fun obtainAllNodeRect(root: DefaultMutableTreeNode) {
        val nodeInfo = root.userObject as NodeInfo
        allNodeRect.add(nodeInfo.bounds)
        for (i in 0 until root.childCount) {
            val child = root.getChildAt(i) as DefaultMutableTreeNode
            obtainAllNodeRect(child)
        }
    }

    fun obtainAttributeMap() {
        val node = hierarchyDoc!!.documentElement.getElementsByTagName("node")
        val map = mutableMapOf<String, MutableMap<String, Int>>()
        for (i in 0 until node.length) {
            val item = node.item(i)
            for (prop in XPathLite.props) {
                val nodeValue = item.attributes.getNamedItem(prop)?.nodeValue
                if (!nodeValue.isNullOrEmpty()) {
                    val propMap = map.getOrPut(prop) { mutableMapOf() }
                    val count = propMap.getOrDefault(nodeValue, 0)
                    propMap[nodeValue] = count + 1
                }
            }
        }
        attributeMap = map
    }

    fun dumpUI() {
        u2.dumpUI(screenShotSaveFile, hierarchySaveFile)
    }

    fun close() {
        u2.destroy()
    }

    fun updateAlias(alias: String) {
        this.alias = alias
        deviceAliasConfig.setAlias(serial, alias)
    }

    fun setAliasByIndex(index: Int) {
        if (alias != "d") return
        alias = if (index > 0) "d${index + 1}" else "d"
        deviceAliasConfig.setAlias(serial, alias)
    }
}
